lib/src/conflicts.rs

```diff
@@ -276,37 +290,80 @@ fn write_conflict_marker(
     }
 }
 
-/// Parse a conflict marker from a line of a file. The conflict marker must have
-/// the correct length (CONFLICT_MARKER_LEN).
-fn parse_conflict_marker(line: &[u8]) -> Option<ConflictMarkerKind> {
+/// Parse a conflict marker from a line of a file. The conflict marker may have
+/// any length (even less than MIN_CONFLICT_MARKER_LEN).
+fn parse_conflict_marker_any_len(line: &[u8]) -> Option<ConflictMarker> {
     let first_byte = *line.first()?;
     let kind = ConflictMarkerKind::parse_byte(first_byte)?;
     let len = line.iter().take_while(|&&b| b == first_byte).count();
 
-    if len != CONFLICT_MARKER_LEN {
-        return None;
-    }
-
     if let Some(next_byte) = line.get(len) {
         // If there is a character after the marker, it must be ASCII whitespace
         if !next_byte.is_ascii_whitespace() {
             return None;
         }
     }
 
-    Some(kind)
+    Some(ConflictMarker { kind, len })
+}
+
+/// Parse a conflict marker, expecting it to have a certain length. Any conflict
+/// markers which have a different length are ignored.
+fn parse_conflict_marker(line: &[u8], expected_len: usize) -> Option<ConflictMarkerKind> {
+    debug_assert!(expected_len >= MIN_CONFLICT_MARKER_LEN);
+
+    parse_conflict_marker_any_len(line)
+        .filter(|marker| marker.len == expected_len)
+        .map(|marker| marker.kind)
+}
+
+/// Find the longest conflict marker in a file.
+pub fn find_max_conflict_marker_len(file: &[u8]) -> usize {
+    file.lines_with_terminator()
+        .filter_map(parse_conflict_marker_any_len)
+        .map(|marker| marker.len)
+        .max()
+        .unwrap_or_default()
 }
 
+/// Choose an appropriate conflict marker length to use for a file. The conflict
+/// marker length will be at least MIN_CONFLICT_MARKER_LENGTH, and is guaranteed
+/// to be longer than any existing markers in the file.
+pub fn choose_conflict_marker_len<T: AsRef<[u8]>>(single_hunk: &Merge<T>) -> usize {
+    let max_existing_marker_len = single_hunk
+        .iter()
+        .map(|term| find_max_conflict_marker_len(term.as_ref()))
+        .max()
+        .unwrap_or_default();
+
+    if max_existing_marker_len < MIN_CONFLICT_MARKER_LEN {
```

Comment:
nit: perhaps, this should be `max(max_existing_marker_len + INCREMENT, MIN_LEN)`?

If existing literal were `<<<<<<`, marker of the length 7 would be hard to distinguish.